design/authenticators/authn_k8s/inject_client_cert_solution_design.md

```diff
@@ -0,0 +1,255 @@
+We don't have backwards-compatible issues with this change as also before this change the server didn't wait for the cert to be injected. The server closed the socket to kubectl before waiting for the cert to be injected. Furthermore, the authn-client just verifies that the response code is less than 300 so changing 200 to 202 doesn't affect it.
+
+### Performance
+[//]: # "Elaborate on whether your solution will affect the product's performance, and how"
+The performance of the "inject-client-cert" request will not be affected as the Websocket is still being closed immediately after we inject the cert.
+
+### Security
+
+[//]: # "Are there any security issues with your solution? Even if you mentioned them somewhere in the doc it may be convenient for the security architect review to have them centralized here"
+
+1. Is it ok to pass the base64 encoded cert in the URL? Is it less secure than sending it in the STDIN?
+   1. The answer is yes and thus we are not going with option 2.
+
+## Affected Components
+
+[//]: # "Address the components that will be affected by your solution [Conjur, DAP, clients, integrations, etc.]"
+- Conjur
+- DAP
+- conjur-authn-k8s-client
+- seed-fetcher
+- secrets-provider-for-k8s
+
+## Cross-team dependencies
+[//]: # "Elaborate about your solution's requirements or needs from other teams"
+We would like to get a review on the solution from @cyberark/conjur-core-team  and @cyberark/community-and-integrations-team. There are no requirements from  the Infra team.
+
+## Test Plan
+[//]: # "Fill in the table below to depict the tests that should run to validate your solution"
+[//]: # "You can use this tool to generate a table - https://www.tablesgenerator.com/markdown_tables#"
+
+The current integration tests of the "inject-client-cert" request are verifying that the cert is installed but they have 10 retries of 2 seconds each. If we decide to make the request atomic (i.e return 200 OK upon cert injection) then we should change this behaviour to have no retries.
+
+| **Given**                                                    | **When**                                | **Then**                                                     | **Comment**                                                  |
+| ------------------------------------------------------------ | --------------------------------------- | ------------------------------------------------------------ | ------------------------------------------------------------ |
+| An authn-client container with permissions on the "ssl" directory | An "inject-client-cert" request is sent | - The response code is 200 OK<br />- The cert is in the container |                                                              |
+| An authn-client container without permissions on the "ssl" directory | An "inject-client-cert" request is sent | - The response code is 401<br />- The error is written to the log |                                                              |
```

Comment:
how would the server determine the response should be 401? is this something that's already implemented? if this is new, it would be good to talk about this a bit more elsewhere

is 401 even the right response in this case, when a valid client sends a request but the directory in the client container doesn't exist or doesn't have the right permissions? the status code definition for 401 says
> specifically for use when authentication is required and has failed or has not yet been provided.

I don't think that matches this case. I'm not sure what the right response would be - maybe 403 Forbidden (`The request contained valid data and was understood by the server, but the server is refusing action`)?

Reply:
you are correct, it's a leftover. it should be 202 in this case.

Reply:
i updated the test plan.